lib/tasks/version_scheme_counter.rake

```diff
@@ -0,0 +1,135 @@
+# 2. The one with the highest tally is likely the scheme which was being followed.
+# 3. If tied for unknown, mark unknown.
+# 4. Since calver validates as other schemes, prefer calver in the case of a tie since it is unusual and likely calver.
+# 5. In the case of a tie, if Semver is one of the contenders, prefer Semver for its ubiquity.
+# 6. Assume only a Maven project would use the Maven or OSGi versioning schemes.
+#
+# Known issues: many version schemes will match PEP440 and even Maven and yet be Semver. In the case of a tie, we choose
+#   semver, but this may turn out to be the wrong choice later on.
+#
+# Note: Maven versioning scheme is a looser form of OSGi
+
+module VersionSchemeDetection
+  REGEXES = {
+    # Based on https://docs.oracle.com/middleware/1212/core/MAVEN/maven_version.htm#MAVEN8855
+    # Derived from https://github.com/mojohaus/build-helper-maven-plugin/blob/master/src/main/java/org/codehaus/mojo/buildhelper/versioning/VersionInformation.java
+    MAVEN: /^\d+(\.\d+)?{1,2}(.*)$/,
+    # OSGI versioning can be expected primarily on Maven projects.
+    # Derived from http://docs.osgi.org/specification/osgi.core/7.0.0/framework.module.html#i2655136
+    OSGI: /^\d+(\.\d+)?{1,2}(\.[a-zA-Z0-9_-]+)?$/,
+    # PEP440 Regex copied from https://www.python.org/dev/peps/pep-0440/#appendix-b-parsing-version-strings-with-regular-expressions
+    PEP440: /^([1-9][0-9]*!)?(0|[1-9][0-9]*)(\.(0|[1-9][0-9]*))*((a|b|rc)(0|[1-9][0-9]*))?(\.post(0|[1-9][0-9]*))?(\.dev(0|[1-9][0-9]*))?$/
+  }
+
+  # validators return truthy/falsy
+  VALIDATORS = {
+    MAVEN: ->(version) { version =~ REGEXES[:MAVEN] },
+    PEP440: ->(version) { version =~ REGEXES[:PEP440] },
+    OSGI: ->(version) { version =~ REGEXES[:OSGI] },
+    SEMVER: ->(version) { SemanticRange.valid(version).present? },
+    CALVER: lambda do |version|
+      first_part = version.split(".")[0]
+      # Assume most if not all Calver versions were released after 1000 AD
+      # Also assume if they are using YY as the first part that it is after 2000 AD
+      first_part.length >= 4 || (first_part.length == 2 && first_part.to_i > 0)
+    end
+  }
+end
+
+tallies = { semver: 0, pep440: 0, maven: 0, osgi: 0, calver: 0, unknown: 0, no_versions: 0 }
+
+namespace :version do
+  desc 'Tests a sampling of project\'s versions to count occurrences of different versioning schemes'
+  task :scheme_counter, [:package_list] => :environment do |t, args|
+    raise "Provide package_list csv file with columns [package_platform, package_name]" unless args[:package_list].present?
+
+    global_tallies = tallies.clone.merge({
+      unknown_schemes: []
+    })
+    warnings = []
+
+    packages = CSV.read(args[:package_list])
+    # downcase platform
+    packages.map! { |package| [package[0].downcase, package[1]] }
+
+    # The following makes a query like "(x AND y) OR (x AND y)"
+    package_matcher = "(LOWER(platform) = ? AND name = ?)"
```

Comment:
There's a more programmatic and less string-oriented way you can build these SQL statements. We use Arel nodes on the Catalog model in DepCI to build a complex grouping of `OR` queries that is a bit easier to work with: https://github.com/tidelift/dependencyci/blob/main/app/models/catalog.rb#L634-L645